src/fontra/views/editor/scene-controller.js

```diff
@@ -1148,6 +1168,49 @@ export class SceneController {
     });
   }
 
+  async closeContour() {
+    const openContours = this.contextMenuState.openContourSelection;
+    const curveType = this.experimentalFeatures.quadPenTool ? "quad" : "cubic";
+    await this.editLayersAndRecordChanges((layerGlyphs) => {
+      for (const layerGlyph of Object.values(layerGlyphs)) {
+        const path = layerGlyph.path;
+        for (const contourIndex of openContours) {
+          // close open contour
+          path.contourInfo[contourIndex].isClosed = true;
+          if (curveType === "cubic") {
+            this.closeContourEnsureCubicOffCurves(path, contourIndex);
+          }
+        }
+      }
+      return "Close Contour" + (openContours.length > 1 ? "s" : "");
+    });
+  }
+
+  closeContourEnsureCubicOffCurves(path, contourIndex) {
+    const contourInfo = path.contourInfo[contourIndex];
+    const numContourPoints = path.getContour(contourIndex).pointTypes.length;
+
+    const prevEndPoint = path.getPoint(contourInfo.endPoint - 1);
+    const endPoint = path.getPoint(contourInfo.endPoint);
+    const startPoint = path.getPoint(contourInfo.endPoint - numContourPoints + 1);
+
+    if (prevEndPoint.type || !endPoint.type || startPoint.type) {
+      // Sanity check: we expect on-curve/off-curve/on-curve
+      return;
+    }
+
+    // Compute handles for a cubic segment that will look the same as the
```

Comment:
This code currently only works if the dangling point is at the end. But it should also work if it's at the start. Reverse the contour then "close contour" should work just the same.

Reply:
Great edge cases. done.